python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java

```diff
@@ -52,9 +55,27 @@ public boolean isIdentityComparableWith(InferredType other) {
     if (other instanceof UnionType) {
       return other.isIdentityComparableWith(this);
     }
+    if (isComparingTypeWithMetaclass(other)) {
+      return true;
+    }
     return this.equals(other);
   }
 
+  private boolean isComparingTypeWithMetaclass(InferredType other) {
+    if (other instanceof RuntimeType otherRuntimeType) {
+      boolean hasOtherMetaClass = hasMetaclassInHierarchy(otherRuntimeType);
+      boolean hasThisMetaClass = hasMetaclassInHierarchy(this);
+      return (TYPE_CLASS_NAME.equals(getTypeClass().name()) && hasOtherMetaClass)
+        || (hasThisMetaClass && TYPE_CLASS_NAME.equals(otherRuntimeType.getTypeClass().name()));
+    }
+    return false;
+  }
+
+  private static boolean hasMetaclassInHierarchy(RuntimeType runtimeType) {
+    return runtimeType.getTypeClass().hasMetaClass()
+      || runtimeType.getTypeClass().superClasses().stream().filter(ClassSymbol.class::isInstance).anyMatch(c -> ((ClassSymbol) c).hasMetaClass());
```

Comment:
should we look only for current type parents for a `metaclass`, or should we go deeper in the types hierarchy?

Reply:
I intend to go deeper as the original issue contained a child class from `enum.Enum`

Reply:
as discussed - we need to check whole hierarchy, not nearest parents